analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -144,6 +132,37 @@ kind switch
             };
     }
 
+    private static NumberConstraint RelationalNumberConstraint(NumberConstraint existingNumber, BinaryOperatorKind kind, NumberConstraint comparedNumber, int visitCount)
+    {
+        return kind switch
+        {
+            BinaryOperatorKind.Equals => NumberConstraint.From(BiggestMinimum(comparedNumber, existingNumber), SmallestMaximum(comparedNumber, existingNumber)),
+            BinaryOperatorKind.NotEquals when comparedNumber.IsSingleValue && comparedNumber.Min == existingNumber?.Min => NumberConstraint.From(existingNumber.Min + 1, existingNumber.Max),
+            BinaryOperatorKind.NotEquals when comparedNumber.IsSingleValue && comparedNumber.Min == existingNumber?.Max => NumberConstraint.From(existingNumber.Min, existingNumber.Max - 1),
+            BinaryOperatorKind.GreaterThan when comparedNumber.Min.HasValue => From(comparedNumber.Min + 1, null),
+            BinaryOperatorKind.GreaterThanOrEqual when comparedNumber.Min.HasValue => From(comparedNumber.Min, null),
+            BinaryOperatorKind.LessThan when comparedNumber.Max.HasValue => From(null, comparedNumber.Max - 1),
+            BinaryOperatorKind.LessThanOrEqual when comparedNumber.Max.HasValue => From(null, comparedNumber.Max),
+            _ => null
+        };
+
+        NumberConstraint From(BigInteger? newMin, BigInteger? newMax)
+        {
+            if (existingNumber is not null)
+            {
+                if (!newMin.HasValue || (existingNumber.Min > newMin && IsFirstLoopVisit(visitCount)))
+                {
+                    newMin = existingNumber.Min;
+                }
+                if (!newMax.HasValue || (existingNumber.Max < newMax && IsFirstLoopVisit(visitCount)))
+                {
+                    newMax = existingNumber.Max;
+                }
+            }
+            return NumberConstraint.From(newMin, newMax);
+        }
```

Comment:
This can be simplified:
```suggestion
        NumberConstraint From(BigInteger? newMin, BigInteger? newMax)
        {
            if (newMin is null || (existingNumber?.Min > newMin && IsFirstLoopVisit(visitCount)))
            {
                newMin = existingNumber?.Min;
            }
            if (newMax is null || (existingNumber?.Max < newMax && IsFirstLoopVisit(visitCount)))
            {
                newMax = existingNumber?.Max;
            }
            return NumberConstraint.From(newMin, newMax);
        }
```

Reply:
I prefer the nested version with pre-check for existingNumber.

You can do the `newMin is null` thing instead of `! .HasValue` in your PR. You'll need to touch this part anyway.